### PR TITLE
Store configuration status and enable remote watch events

### PR DESCRIPTION
**Pull request by JohnMcPMS**
## Change
This change stores ongoing status information during configuration application in the database.  This data is then retrievable via the existing API surface from a set retrieved from `GetConfigurationHistory`.  The data includes: configuration set state, configuration set apply start/stop times, configuration unit state, and configuration unit results.

In addition, the existing events on the `ConfigurationProcessor` and `ConfigurationSet` are implemented to send data when appropriate.  This works across processes, using named events to indicate changes and the database to carry the actual data.

`winget configure list` is updated to expose this information, for example:

```
PS > wingetdev configure list
Identifier                             Name      State     Origin
--------------------------------------------------------------------
{7D5CF50E-F3C6-4333-BFE6-5A806F9EBA4E} Test Name Completed Test Path
{51837FDE-C290-408B-A847-B176DB184B89} Test Name Completed Test Path
{E7367B3D-07A2-4875-A190-C369B79E6F5B} Test Name Completed Test Path
{B186F2D0-3245-4AD1-A712-260906033E57} Test Name Completed Test Path
{2AA76541-A0DF-45EE-91F4-401F93FA88CD} Test Name Completed Test Path
PS > wingetdev configure list --history 7D5CF50E
Field         Value
----------------------------------------------------
Identifier    {7D5CF50E-F3C6-4333-BFE6-5A806F9EBA4E}
Name          Test Name
Origin        Test Origin
Path          Test Path
State         Completed
First Applied 2024-07-16 21:15:13.000
Apply Begun   2024-07-16 21:15:13.000
Apply Ended   2024-07-16 21:15:13.000

Unit                         State     Result
-------------------------------------------------
Module/Resource [Name]       Completed 0x00000000
Module/Resource2 [Name2]     Completed 0x00000000
Module2/Resource [Group]     Completed 0x00000000
|-Module3/Resource [Child1]  Completed 0x00000000
|-Module4/Resource2 [Child2] Completed 0x00000000
```
 
## Validation
Updated tests to check for status data.  Manually verified eventing.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoft/winget-cli/pull/4647)

### CODE REVIEW

**Review by florelis — APPROVED** [review]
lgtm, but I'll admit that a lot of the nuance was lost on me...

File: src/AppInstallerCLICore/Workflows/ConfigurationFlow.cpp [hunk 1]
```diff
@@ -1897,18 +1964,108 @@ namespace AppInstaller::CLI::Workflow
...
+            }
+
+            ConfigurationUnit& currentUnit = currentSiblings.Siblings[currentSiblings.Current++];
+
+            std::ostringstream unitStream;
+
+            if (currentSiblings.Depth)
+            {
+                unitStream << '|' << std::string((currentSiblings.Depth * 2) - 1, '-');
+            }
```
[the remark this answers is not in the document]
Comment (JohnMcPMS) [reply src/AppInstallerCLICore/Workflows/ConfigurationFlow.cpp:2028]: Yes, the goal was that a flat list would not need any additional characters, while child items would hopefully be clearly indicated.

File: src/AppInstallerSharedLib/SQLiteStatementBuilder.cpp [hunk 2]
```diff
@@ -288,6 +288,20 @@ namespace AppInstaller::SQLite::Builder
...
+    {
+        OutputColumns(m_stream, "MAX(", column);
+        m_stream << ")";
+        return *this;
+    }
+
+    StatementBuilder& StatementBuilder::Max(const QualifiedColumn& column)
+    {
+        OutputColumns(m_stream, "MAX(", column);
+        m_stream << ")";
```
[the remark this answers is not in the document]
Comment (JohnMcPMS) [reply src/AppInstallerSharedLib/SQLiteStatementBuilder.cpp:301]: So this is why I thought it was weird there wasn't already a method... there was.

File: src/Microsoft.Management.Configuration/Database/ConfigurationDatabase.cpp
[the remark this answers is not in the document]
Comment (JohnMcPMS) [reply src/Microsoft.Management.Configuration/Database/ConfigurationDatabase.cpp:1]: Used some template methods to remove ~250 lines of repetitive code.

**Review by florelis — APPROVED** [review]

**Pull request merged**